File: client-app/src/models/budgetItem.ts
export class BudgetItem {
    id: string;
    title: string;
    target: number;
    outflow: number;
    assigned: number;
    budgetGroupId: string;
    dateCreated: string;

    constructor(id: string, name: string, categoryGroupId: string, dateCreated: string, assigned: number = 0, target: number = 0, outflow: number = 0) {
        this.id = id;
        this.title = name;
        this.assigned = assigned;
        this.target = target;
        this.outflow = outflow;
        this.budgetGroupId = categoryGroupId;
        this.dateCreated = dateCreated;
    }

    private calculateAvailable(): number {
        return this.assigned - (-this.outflow);
    }

    // Getter for available
    get available(): number {
        return this.calculateAvailable();
    }

    static fromBudgetItem(budgetItem: BudgetItem): BudgetItem {
        return new BudgetItem(
            budgetItem.id,
            budgetItem.title,
            budgetItem.budgetGroupId,
            budgetItem.dateCreated,
            budgetItem.assigned,
            budgetItem.target,
            budgetItem.outflow
        );
    }
}

File: Domain/Models/Budgets/BudgetItem.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Domain.Models.Budgets;

/// <summary>
/// Represents a category. This is an individual budget item
/// such as I want to save 200 a bucks a month, or have
/// a budget of $600 a month for groceries.
/// </summary>
public class BudgetItem
{
    /// <summary>
    /// The <see cref="BudgetItem.Id"/>
    /// </summary>
    [Required]
    public Guid Id { get; set; }

    /// <summary>
    /// The <see cref="BudgetItem.Title"/>
    /// </summary>
    public required string Title { get; set; }

    /// <summary>
    /// How much to assign to this <see cref="BudgetItem"/>
    /// </summary>
    public double Target { get; set; }

    /// <summary>
    /// How much has been spent on this <see cref="BudgetItem"/>
    /// </summary>
    [NotMapped]
    public double Outflow { get; set; }

    /// <summary>
    /// How much has been assigned to this <see cref="BudgetItem"/>
    /// </summary>
    [NotMapped]
    public double Assigned { get; set; }

    /// <summary>
    /// the <see cref="BudgetGroup.Id"/> of this <see cref="BudgetItem"/>
    /// </summary>
    public Guid BudgetGroupId { get; set; }

    /// <summary>
    /// The date when
    /// </summary>
    public DateTime DateCreated { get; set; }
}